languages/go/types/polar_types.go

```diff
@@ -15,6 +15,593 @@ type Call struct {
+	var variantValue *json.RawMessage
+
+	// try and deserialize as a string first
+	err := json.Unmarshal(b, &variantName)
+	if err != nil {
+		var rawMap map[string]json.RawMessage
+		err := json.Unmarshal(b, &rawMap)
+		if err != nil {
+			return err
+		}
+		// JSON should be of form {"VariantName": {...}}
+		if len(rawMap) != 1 {
+			return errors.New("deserializing Comparison as an enum variant; expecting a single key")
+		}
+		for k, v := range rawMap {
+			variantName = k
+			variantValue = &v
+		}
+	}
+	switch variantName {
+	case "Eq":
+		var variant ComparisonEq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Neq":
+		var variant ComparisonNeq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "In":
+		var variant ComparisonIn
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Nin":
+		var variant ComparisonNin
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Lt":
+		var variant ComparisonLt
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Leq":
+		var variant ComparisonLeq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Gt":
+		var variant ComparisonGt
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	case "Geq":
+		var variant ComparisonGeq
+		if variantValue != nil {
+			err := json.Unmarshal(*variantValue, &variant)
+			if err != nil {
+				return err
+			}
+		}
+		*result = Comparison{variant}
+		return nil
+	}
+
+	return fmt.Errorf("cannot deserialize Comparison: %s", string(b))
+}
+
+func (variant Comparison) MarshalJSON() ([]byte, error) {
+	switch inner := variant.ComparisonVariant.(type) {
+	case ComparisonEq:
+		return json.Marshal("Eq")
+	case ComparisonNeq:
+		return json.Marshal("Neq")
+	case ComparisonIn:
+		return json.Marshal("In")
+	case ComparisonNin:
+		return json.Marshal("Nin")
+	case ComparisonLt:
+		return json.Marshal("Lt")
+	case ComparisonLeq:
+		return json.Marshal("Leq")
+	case ComparisonGt:
+		return json.Marshal("Gt")
+	case ComparisonGeq:
+		return json.Marshal("Geq")
+	default:
+		return nil, fmt.Errorf("unexpected variant %#v of %v", inner, variant)
+	}
+
+}
+
+// Condition struct
+type Condition struct {
```

Comment:
This is what tuples look like on the Go side. It's not nice. Would be great if these were structs instead

Reply:
Yeah, we wanted to make that change anyway. I had parsed them into go structs in my version I think.